### PR TITLE
feat: allow maxCallRecvMsgSize to be increased for large secret mounts

### LINKED ISSUE #460: Proposal: secret-store-csi-driver file IO consolidation
**Describe the solution you'd like**
This is a tracking issue for the [secret-store-csi-driver file IO consolidation proposal](https://docs.google.com/document/d/1y1Oc1_-JGxMh-MZ57iqrVnoCIn6V-W0vJcj3qmq8EWU/edit#).

**Anything else you would like to add:**

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by tam7t**
**What this PR does / why we need it**:

In https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/481#discussion_r601917265 secret contents are returned through the grpc interface, but grpc has a max-message size (4MiB).

From [discussion on the proposal](https://docs.google.com/document/d/1y1Oc1_-JGxMh-MZ57iqrVnoCIn6V-W0vJcj3qmq8EWU/edit?disco=AAAALghdMrs) this PR introduces a way to increase the max message size in the event that a user needs a very large mount.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #460

**Special notes for your reviewer**:

### CODE REVIEW

File: cmd/secrets-store-csi-driver/main.go [hunk 1]
```diff
@@ -58,6 +59,7 @@ var (
 	rotationPollInterval = flag.Duration("rotation-poll-interval", 2*time.Minute, "Secret rotation poll interval duration")
 	enableProfile        = flag.Bool("enable-pprof", false, "enable pprof profiling")
 	profilePort          = flag.Int("pprof-port", 6065, "port for pprof profiling")
+	maxCallRecvMsgSize   = flag.Int("max-call-recv-msg-size", 1024*1024*4, "maximum size of grpc response from plugins")
```
Comment (tomhjp) [inline cmd/secrets-store-csi-driver/main.go:62]:
```suggestion
	maxCallRecvMsgSize   = flag.Int("max-call-recv-msg-size", 1024*1024*4, "maximum size in bytes of gRPC response from plugins")
```

File: pkg/secrets-store/provider_client.go [hunk 2]
```diff
@@ -82,12 +83,23 @@ type PluginClientBuilder struct {
 // 		<path>/<plugin_name>.sock
 //
 // where <plugin_name> must match the PluginNameRe regular expression.
-func NewPluginClientBuilder(path string) *PluginClientBuilder {
+//
+// Additional grpc dial options can also be set through opts and will be when
+// creating all clients.
```
Comment (tomhjp) [inline pkg/secrets-store/provider_client.go:88]:
```suggestion
// Additional grpc dial options can also be set through opts and will be used when
// creating all clients.
```

File: pkg/secrets-store/provider_client_test.go [hunk 3]
```diff
@@ -157,6 +160,45 @@ func TestMountContent(t *testing.T) {
 	}
 }
 
+func TestMountContent_TooLarge(t *testing.T) {
```
Comment (tomhjp) [inline pkg/secrets-store/provider_client_test.go:163]: Nice test!

**Review by tomhjp — APPROVED** [review]
LGTM! I'm ok if your answer is no, but do you think it's worth adding a note to troubleshooting, or anywhere else in the docs?

File: cmd/secrets-store-csi-driver/main.go (hunk 1 above)
Comment (tam7t) [reply cmd/secrets-store-csi-driver/main.go]: done thanks!

File: pkg/secrets-store/provider_client.go (hunk 2 above)
Comment (tam7t) [reply pkg/secrets-store/provider_client.go]: done thanks!

File: pkg/secrets-store/provider_client_test.go (hunk 3 above)
Comment (tam7t) [reply pkg/secrets-store/provider_client_test.go:163]: thanks!

**Comment by tam7t** [comment]
I added a note to the troubleshooting page so that its at least mentioned

File: cmd/secrets-store-csi-driver/main.go [hunk 4]
```diff
@@ -24,6 +24,7 @@ import (
 	_ "net/http/pprof" // #nosec
 	"time"
 
+	"google.golang.org/grpc"
```
Comment (aramase) [inline cmd/secrets-store-csi-driver/main.go:27]: nit: we should group the imports in the following order

```
import (
  stdlib

  internal packages

  external packages
)
```

We can do this cleanup in a follow up PR for all existing packages

File: docs/book/src/troubleshooting.md [hunk 5]
```diff
@@ -53,4 +53,13 @@ Secrets Store CSI Driver is deployed as a *DaemonSet*. The above error indicates
   NAME                       ATTACHREQUIRED   PODINFOONMOUNT   MODES       AGE
   secrets-store.csi.k8s.io   false            true             Ephemeral   110m
   ```
-  
\ No newline at end of file
+
+### Mount fails with `grpc: received message larger than max`
+
+The files pulled in by the `SecretProviderClass` are larger than 4MiB you may observe `FailedMount` warnings with a
```
Comment (aramase) [inline docs/book/src/troubleshooting.md:59]: This sentence doesn't seem to be complete. 

If the files pulled in by the `SecretProviderClass` are larger than 4MiB you may observe `FailedMount` warnings with a message that include `grpc: received message larger than max`.

**Review by aramase — COMMENTED** [review]
Should we also add this new flag to the helm charts?

File: cmd/secrets-store-csi-driver/main.go (hunk 4 above)
Comment (tam7t) [reply cmd/secrets-store-csi-driver/main.go]: 👍 agreed, I moved the grpc one but didnt move the others cause im not quite sure about `kubebuilder:scaffold:imports` as my goimports config moves some things around weirdly

File: docs/book/src/troubleshooting.md (hunk 5 above)
Comment (tam7t) [reply docs/book/src/troubleshooting.md]: 😢 fixed

**Review by tam7t — COMMENTED** [review]
added helm

**Review by aramase — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**